File: macro-dashboard/backend/app/schemas/replay.py
"""
Replay Lab schemas — historical playbook replay, outcome review, and context notes.

No catalysts field: build_catalyst_state() is not date-aware and is intentionally
omitted from replay responses. Curated context notes serve as the historical
forward-watch layer.
"""

from __future__ import annotations

from pydantic import BaseModel, Field

from app.schemas.dashboard_state import DashboardState
from app.schemas.indicator_snapshot import IndicatorSnapshot
from app.schemas.playbook_conclusion import PlaybookConclusion
from app.schemas.source_meta import SourceMeta
from app.schemas.summary import PlaybookSummary


class ContextNote(BaseModel):
    """A curated historical context note covering a date range."""
    title: str
    body: str
    tags: list[str] = Field(default_factory=list)


class OutcomePoint(BaseModel):
    """Price outcomes for a single forward time horizon."""
    date_end: str           # ISO date of the horizon endpoint
    # SPY (SPDR S&P 500 ETF) — field names match the ticker used
    spy_return_pct: float | None = None
    spy_price_start: float | None = None
    spy_price_end: float | None = None
    # QQQ (Invesco Nasdaq-100 ETF)
    qqq_return_pct: float | None = None
    qqq_price_start: float | None = None
    qqq_price_end: float | None = None
    # WTI crude (CL=F) — closes the inflation/oil loop for the user
    wti_change_pct: float | None = None
    wti_price_start: float | None = None
    wti_price_end: float | None = None


class OutcomeReview(BaseModel):
    """Forward price outcomes at 1W, 1M, and 3M from the replay date."""
    as_of: str
    outcomes_1w: OutcomePoint
    outcomes_1m: OutcomePoint
    outcomes_3m: OutcomePoint
    data_note: str = ""


class ReplayPlaybookResponse(BaseModel):
    """
    Full replay response for a historical as_of date.

    Intentionally omits `catalysts` — build_catalyst_state() relies on live
    config and is not date-aware. The `context` list (curated notes) serves
    as the historical forward-watch layer instead.
    """
    as_of: str
    snapshot: IndicatorSnapshot
    state: DashboardState
    playbook_conclusion: PlaybookConclusion | None = None
    summary: PlaybookSummary
    sources: dict[str, SourceMeta] = Field(default_factory=dict)
    context: list[ContextNote] = Field(default_factory=list)
    data_notes: list[str] = Field(default_factory=list)
    generated_at: str

File: macro-dashboard/frontend/src/types/replay.ts
// Replay Lab types — mirror backend replay.py schemas exactly.
// Note: no `catalysts` field — build_catalyst_state() is not date-aware
// and is intentionally omitted from replay responses.

import type { DashboardState, PlaybookSummary } from './summary'
import type { IndicatorSnapshot } from './indicator'
import type { PlaybookConclusion, SourceMeta } from './playbook'

export interface ContextNote {
  title: string
  body: string
  tags: string[]
}

export interface OutcomePoint {
  date_end: string
  // SPY (SPDR S&P 500 ETF)
  spy_return_pct: number | null
  spy_price_start: number | null
  spy_price_end: number | null
  // QQQ (Invesco Nasdaq-100 ETF)
  qqq_return_pct: number | null
  qqq_price_start: number | null
  qqq_price_end: number | null
  // WTI crude (CL=F)
  wti_change_pct: number | null
  wti_price_start: number | null
  wti_price_end: number | null
}

export interface OutcomeReview {
  as_of: string
  outcomes_1w: OutcomePoint
  outcomes_1m: OutcomePoint
  outcomes_3m: OutcomePoint
  data_note: string
}

export interface ReplayPlaybookResponse {
  as_of: string
  snapshot: IndicatorSnapshot
  state: DashboardState
  playbook_conclusion?: PlaybookConclusion
  summary: PlaybookSummary
  sources: Record<string, SourceMeta>
  context: ContextNote[]
  data_notes: string[]
  generated_at: string
}

// Local-only type — stored in localStorage, never sent to the server
export interface ReplayAssessment {
  date: string                            // YYYY-MM-DD
  regime: string
  posture: string
  watchpoints: [string, string, string]
  confidence: 'High' | 'Medium' | 'Low'
  notes: string
  saved_at: string                        // ISO timestamp
  revealed: boolean
}
